ingress/controllers/nginx/controller.go

```diff
@@ -217,6 +248,126 @@ func (lbc *loadBalancerController) getUDPConfigMap(ns, name string) (*api.Config
+	}
+
+	if !ingExists {
+		glog.Warningf("service %v not found", key)
+		return
+	}
+
+	ing := ingObj.(*extensions.Ingress)
+	for _, rule := range ing.Spec.Rules {
+		if rule.IngressRuleValue.HTTP == nil {
+			continue
+		}
+
+		for _, path := range rule.HTTP.Paths {
+			svcKey := fmt.Sprintf("%v/%v", ing.GetNamespace(), path.Backend.ServiceName)
+			svcObj, svcExists, err := lbc.svcLister.Store.GetByKey(svcKey)
+			if err != nil {
+				glog.Infof("error getting service %v from the cache: %v", svcKey, err)
+				continue
+			}
+
+			if !svcExists {
+				glog.Warningf("service %v does no exists", svcKey)
+				continue
+			}
+
+			svc := svcObj.(*api.Service)
+			if svc.Spec.Selector == nil {
+				return
+			}
+
+			err = lbc.checkSvcForUpdate(svc)
+			if err != nil {
+				lbc.svcEpQueue.requeue(key, err)
+				return
+			}
+		}
+	}
+}
+
+// checkSvcForUpdate verifies if one of the running pods for a service contains
+// named port. If the annotation in the service does not exists or is not equals
+// to the port mapping obtained from the pod the service must be updated to reflect
+// the current state
+func (lbc *loadBalancerController) checkSvcForUpdate(svc *api.Service) error {
+	// get the pods associated with the service
+	pods, err := lbc.client.Pods(svc.Namespace).List(api.ListOptions{
+		LabelSelector: labels.Set(svc.Spec.Selector).AsSelector(),
+	})
+	if err != nil {
+		return fmt.Errorf("error searching service pods %v/%v: %v", svc.Namespace, svc.Name, err)
+	}
+
+	if len(pods.Items) == 0 {
+		return nil
+	}
+
+	namedPorts := map[string]string{}
+	// we need to check only one pod searching for named ports
+	pod := &pods.Items[0]
+	glog.V(4).Infof("checking pod %v/%v for named port information", pod.Namespace, pod.Name)
+	for i := range svc.Spec.Ports {
+		servicePort := &svc.Spec.Ports[i]
+
+		_, err := strconv.Atoi(servicePort.TargetPort.StrVal)
+		if err != nil {
+			portNum, err := podutil.FindPort(pod, servicePort)
+			if err != nil {
+				glog.V(4).Infof("failed to find port for service %s/%s: %v", svc.Namespace, svc.Name, err)
+				continue
+			}
+
+			if servicePort.TargetPort.StrVal == "" {
+				continue
+			}
+
+			namedPorts[servicePort.TargetPort.StrVal] = fmt.Sprintf("%v", portNum)
+		}
+	}
+
+	if svc.ObjectMeta.Annotations == nil {
+		svc.ObjectMeta.Annotations = map[string]string{}
+	}
+
+	curNamedPort := svc.ObjectMeta.Annotations[namedPortAnnotation]
+	if !reflect.DeepEqual(curNamedPort, namedPorts) {
+		data, _ := json.Marshal(namedPorts)
+
+		newSvc, err := lbc.client.Services(svc.Namespace).Get(svc.Name)
+		if err != nil {
+			return fmt.Errorf("error getting service %v/%v: %v", svc.Namespace, svc.Name, err)
+		}
+
+		if svc.ObjectMeta.Annotations == nil {
+			svc.ObjectMeta.Annotations = map[string]string{}
+		}
+
+		newSvc.ObjectMeta.Annotations[namedPortAnnotation] = string(data)
+		glog.Infof("updating service %v with new named port mappings", svc.Name)
+		_, err = lbc.client.Services(svc.Namespace).Update(svc)
```

Comment:
That's icky, can we make it more straightforward? 
1. Only trigger searching through pods if there is a named port
2. Return the matching endpoint instead of waiting for out of band watch to fire